functions.php

```diff
@@ -0,0 +1,270 @@
+    }
+
+    return $products[$category];
+}
+
+function addToList(array &$products, array $requestData, ?array $oldProduct = null, bool $newId = false): void
+{
+    $fields = [ID, NAME, PRICE, QUANTITY];
+
+    foreach ($fields as $property) {
+        $updatedProduct[$property] = (
+            isset($requestData[$property]) ? $requestData[$property] :
+            (isset($oldProduct[$property]) ? $oldProduct[$property] : null)
+        );
+    }
+
+    if (!isset($products[$requestData[CATEGORY]])) {
+        $products[$requestData[CATEGORY]] = [];
+    }
+
+    if ($newId) {
+        $updatedProduct[ID] = generateId($products[$requestData[CATEGORY]], ID);
+        $products[$requestData[CATEGORY]][] = $updatedProduct;
+    } else {
+        foreach($products[$requestData[CATEGORY]] as &$product) {
+            if ($product[ID] === $updatedProduct[ID]) {
+                $product = $updatedProduct;
+            }
+        }
+    }
+}
+
+function generateId(array $categoryOfProducts, string $idKeyName): int
+{
+    $id = 0;
+
+    if (count($categoryOfProducts) > 0) {
+        foreach ($categoryOfProducts as $product) {
+            if ($product[$idKeyName] >= $id) {
+                $id = $product[$idKeyName] + 1;
+            }
+        }
+    }
+
+    return $id;
+}
+
+function validateRequestData(array $mandatoryFields, ?array $requestData = null): void
+{
+    if ($requestData === null) {
+        sendResponse(toJson(['error' => 'Sent data is not correct']), 414);
+    }
+
+    $missingFields = array_keys(array_diff_key($mandatoryFields, $requestData));
+
+    if ($missingFields) {
+        sendResponse(toJson(['error' => 'Missing fields: ' . implode(', ', $missingFields)]), 400);
+    }
+
+    $typeErrors = validateType($mandatoryFields, $requestData);
+
+    if($typeErrors) {
+        sendResponse(toJson(['error' => implode(', ', $typeErrors)]), 500);
+    }
+}
+
+function validateType(array $rules, array $requestData, bool $optional = false): ?array
+{
+    $errors = [];
+
+    foreach ($rules as $key => $value) {
+        switch ($value) {
+            case 'int':
+                if ($optional && isset($requestData[$key]) && !is_int($requestData[$key])) {
+                    $errors[] = "Key {$key} must be {$value}";
+                } elseif (!$optional && !is_int($requestData[$key])) {
+                    $errors[] = "Key {$key} must be {$value}";
+                }
+                break;
+            case 'string':
+                if ($optional && isset($requestData[$key]) && !is_string($requestData[$key])) {
+                    $errors[] = "Key {$key} must be {$value}";
+                } elseif (!$optional && !is_string($requestData[$key])) {
+                    $errors[] = "Key {$key} must be {$value}";
+                }
+                break;
+            default:
+                $errors[] = "Can not check {$key} type";
+                break;
+        }
+    }
+
+    return $errors ? $errors : null;
```

Comment:
I see no difference between `[]` and `null` for case when there are no errors.

Reply:
@AndreiOrmanji  - Supposed to return `$errors` in case it's an empty array or use a null coalescing operator ?

Reply:
@AndreiOrmanji  - If there are no differences then will remove the ternary operator and return `$errors` array instead.

Reply:
`[]` can represent "no errors" case. I suggest to return array or errors in any case.

Reply:
@AndreiOrmanji - I used your suggestion in this case.